cosmo_tester/framework/test_hosts.py

```diff
@@ -1080,40 +1101,57 @@ class DistributedInstallationCloudifyManager(TestHosts):
     ROOT_CERT_NAME = 'root.crt'
     ROOT_KEY_NAME = 'root.key'
 
-    def __init__(self, cluster=False, sanity=False, *args, **kwargs):
+    def __init__(self, cluster=False, sanity=False, cluster_upgrade=False,
+                 cluster_version=None, *args, **kwargs):
         self.cluster = cluster
         self.sanity = sanity
 
+        # The environment to upgrade to
+        database_to_use = IMAGES[cluster_version + '_DB'] if cluster_upgrade \
+            else CURRENT_DISTRIBUTED_DATABASE
+        message_queue_to_use = IMAGES[cluster_version + '_RMQ'] if \
+            cluster_upgrade else CURRENT_DISTRIBUTED_MESSAGE_QUEUE
+        manager_to_use = IMAGES[cluster_version + '_MGR_WRKR'] if \
+            cluster_upgrade else CURRENT_DISTRIBUTED_MANAGER
+
         instances = [
-            CURRENT_DISTRIBUTED_DATABASE(),
-            CURRENT_DISTRIBUTED_MESSAGE_QUEUE(),
-            CURRENT_DISTRIBUTED_MANAGER(upload_plugins=True)
+            database_to_use(),
+            message_queue_to_use(),
+            manager_to_use()
         ]
         if cluster:
             instances += [
-                CURRENT_DISTRIBUTED_MANAGER(
-                    upload_plugins=False),
-                CURRENT_DISTRIBUTED_MANAGER(
-                    upload_plugins=False)
+                manager_to_use(upload_plugins=False),
+                manager_to_use(upload_plugins=False)
             ]
             if sanity:
                 instances += [
-                    CURRENT_MANAGER(upload_plugins=True)
+                    CURRENT_MANAGER()
+                ]
+            if cluster_upgrade:
+                # The environment to upgrade from
+                instances += [
+                    IMAGES['4.6'](),
+                    IMAGES['4.6_CLUSTER_JOINED'](upload_plugins=False),
```

Comment:
I think this should be a parameter of test no?

Reply:
This cannot change as it will always be the version we upgrade from, since this is where the we had an old cluster.
So it simulates an old cluster to new cluster upgrade